batchproducer/batchproducer.go

```diff
@@ -187,15 +200,21 @@ func (b *batchProducer) Add(data []byte, partitionKey string) error {
 
 // from/for interface Producer
 func (b *batchProducer) Start() error {
-	if b.isRunning() {
-		return nil
+	b.runningMu.Lock()
+	defer b.runningMu.Unlock()
```

Comment:
It does look ugly. That's why atomics are a much cleaner (and lower-contention) solution to this problem.